### DISCUSSION TITLE
best practice for private local common modules?

### DISCUSSION THREAD

**Discussion by hanguokai** [denoland/deno#13386] [General]
## Use Case

a **local private** common utils module that used by multiple independent other projects.
Currently, I think of two ways: 1. import localhost URL. 2. import files

## import localhost URL
1. start a local file-server at common-module's parent directory
2. `import {...} from "http://localhost:port/a-common-module/mod.ts"`
3. `deno run -r main.ts`

Note:
1. In this way, different projects use the **same** import URL.
2. Because deno run will cache remote(including localhost) modules, to keep the common module update, you needs to use the --reload(-r) option to run in other projects.

## import file directly
For example, `import {...} from "../../some-path/all-local-modules/a-module"`.

But local projects in different locations need to use **different** relative import path.

## Question
Is this the best practice for private local common modules?

**Comment by jsejcksn** [comment]
An [import map](https://deno.land/manual@v1.18.0/linking_to_external_code/import_maps) is an additional tool which can address the "issue" of "wanting to use abbreviated, non-relative specifiers in place of relative URLs".

<details>
<summary>Example:</summary>

`../../common-utils/add/mod.ts`:
```ts
export function add (numbers: number[]): number {
  return numbers.reduce((sum, n) => sum + n);
}
```

`./import_map.json`:
```json
{
  "imports": {
    "common/": "../../common-utils/"
  }
}
```

`./mod.ts`:
```ts
import {add} from 'common/add/mod.ts';

export function average (numbers: number[]): number {
  return add(numbers) / numbers.length;
}
```

`./mod.test.ts`:
```ts
import {assertStrictEquals} from 'https://deno.land/std@0.122.0/testing/asserts.ts';
import {average} from './mod.ts';

Deno.test('averages integers', () => {
  const numbers = [1, 2, 3, 4];
  const actual = average(numbers);
  assertStrictEquals(actual, 2.5);
});
```

```
average % deno --version
deno 1.18.0 (release, x86_64-apple-darwin)
v8 9.8.177.6
typescript 4.5.2

average % deno test --import-map import_map.json
running 1 test from file:///Users/deno/example/projects/average/mod.test.ts
test averages integers ... ok (5ms)

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (22ms)
```
</details>

**Reply by hanguokai** [reply]
Thanks, "import map" is another option for this. Essentially, it moves hard code URL or file path from source code(or deps.ts) to a JSON config file.
